In [1]:
import pandas as pd
import tensorflow as tf
import numpy as np
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.wrappers.scikit_learn import KerasClassifier
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import StratifiedKFold
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline

In [2]:
# Review dataset.
sonar_dataset = pd.read_csv('./sonar.csv',header=None)

In [3]:
sonar_dataset

0       1       2       3       4       5       6       7       8   \
0    0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1    0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2    0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3    0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4    0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   
..      ...     ...     ...     ...     ...     ...     ...     ...     ...   
203  0.0187  0.0346  0.0168  0.0177  0.0393  0.1630  0.2028  0.1694  0.2328   
204  0.0323  0.0101  0.0298  0.0564  0.0760  0.0958  0.0990  0.1018  0.1030   
205  0.0522  0.0437  0.0180  0.0292  0.0351  0.1171  0.1257  0.1178  0.1258   
206  0.0303  0.0353  0.0490  0.0608  0.0167  0.1354  0.1465  0.1123  0.1945   
207  0.0260  0.0363  0.0136  0.0272  0.0214  0.0338  0.0655  0.1400  0.1843   

         9   ...      51      52      53      54      55      56      57  \
0    0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084   
1    0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
2    0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   
3    0.1264  ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044   
4    0.4459  ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048   
..      ...  ...     ...     ...     ...     ...     ...     ...     ...   
203  0.2684  ...  0.0116  0.0098  0.0199  0.0033  0.0101  0.0065  0.0115   
204  0.2154  ...  0.0061  0.0093  0.0135  0.0063  0.0063  0.0034  0.0032   
205  0.2529  ...  0.0160  0.0029  0.0051  0.0062  0.0089  0.0140  0.0138   
206  0.2354  ...  0.0086  0.0046  0.0126  0.0036  0.0035  0.0034  0.0079   
207  0.2354  ...  0.0146  0.0129  0.0047  0.0039  0.0061  0.0040  0.0036   

         58      59  60  
0    0.0090  0.0032   R  
1    0.0052  0.0044   R  
2    0.0095  0.0078   R  
3    0.0040  0.0117   R  
4    0.0107  0.0094   R  
..      ...     ...  ..  
203  0.0193  0.0157   M  
204  0.0062  0.0067   M  
205  0.0077  0.0031   M  
206  0.0036  0.0048   M  
207  0.0061  0.0115   M  

[208 rows x 61 columns]

In [4]:
sonar_dataset[0]

0      0.0200
1      0.0453
2      0.0262
3      0.0100
4      0.0762
        ...  
203    0.0187
204    0.0323
205    0.0522
206    0.0303
207    0.0260
Name: 0, Length: 208, dtype: float64

In [5]:
sonar_dataset.head()

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      51      52      53      54      55      56      57  \
0  0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084   
1  0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
2  0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   
3  0.1264  ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044   
4  0.4459  ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048   

       58      59  60  
0  0.0090  0.0032   R  
1  0.0052  0.0044   R  
2  0.0095  0.0078   R  
3  0.0040  0.0117   R  
4  0.0107  0.0094   R  

[5 rows x 61 columns]

In [6]:
len(sonar_dataset)

208

In [7]:
sonar_dataset.shape

(208, 61)

## Loading Dataset

In [8]:
def train_test_split(data, train_size):
    train = data[:train_size]
    test = data[train_size:]
    return train, test

In [9]:
# fix random seed for reproducibility
seed = 7
np.random.seed(seed)

In [10]:
sonar_dataset_values = sonar_dataset.values
sonar_dataset_values

array([[0.02, 0.0371, 0.0428, ..., 0.009, 0.0032, 'R'],
       [0.0453, 0.0523, 0.0843, ..., 0.0052, 0.0044, 'R'],
       [0.0262, 0.0582, 0.1099, ..., 0.0095, 0.0078, 'R'],
       ...,
       [0.0522, 0.0437, 0.018, ..., 0.0077, 0.0031, 'M'],
       [0.0303, 0.0353, 0.049, ..., 0.0036, 0.0048, 'M'],
       [0.026, 0.0363, 0.0136, ..., 0.0061, 0.0115, 'M']], dtype=object)

In [11]:
# split into input (X) and output (Y) variables
x = sonar_dataset.iloc[:,0:59].astype(float)
y = sonar_dataset.iloc[:,60]
X = sonar_dataset.iloc[:,0:60].astype(float)
Y = sonar_dataset.iloc[:,60]

In [12]:
type(X[0][0])

numpy.float64

In [13]:
encoder = LabelEncoder()
encoded_Y = encoder.fit(Y).transform(Y)
encoded_y = encoder.fit(y).transform(y)
encoded_Y

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [14]:
def create_baseline():
    model = Sequential()
    model.add(Dense(60,activation='relu',input_shape=(60,)))
    model.add(Dense(60,activation='relu'))
    model.add(Dense(1,activation='sigmoid'))
    # Compile.
    model.compile(optimizer='adam',loss='binary_crossentropy',metrics=['acc'])
    return model

In [15]:
estimator = KerasClassifier(build_fn=create_baseline, epochs=100, batch_size=5, verbose=0)
kfold = StratifiedKFold(n_splits=10, shuffle=True, random_state=seed)
results = cross_val_score(estimator, X, encoded_Y, cv=kfold)
print("Results: %.2f%% (%.2f%%)" % (results.mean()*100, results.std()*100))

Instructions for updating:
If using Keras pass *_constraint arguments to layers.
Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where
Results: 84.12% (8.98%)


## Step #3

In [16]:
# evaluate baseline model with standardized dataset
np.random.seed(seed)
estimators = []
estimators.append(('standardize', StandardScaler()))
estimators.append(('mlp', KerasClassifier(build_fn=create_baseline, epochs=100, batch_size=5, verbose=0)))
pipeline = Pipeline(estimators)
kfold = StratifiedKFold(n_splits=10, shuffle=True, random_state=seed)
results = cross_val_score(pipeline, X, encoded_Y, cv=kfold)
print("Standardized: %.2f%% (%.2f%%)" % (results.mean()*100, results.std()*100))

Standardized: 85.55% (6.88%)


## Step #4. Tuning Layers and Number of Neurons in The Model.

### Step #4.1

In [17]:
# smaller model
def create_smaller():
    model = Sequential()
    model.add(Dense(60,activation='relu',input_shape=(60,)))
    model.add(Dense(30,activation='relu'))
    model.add(Dense(1,activation='sigmoid'))
    # Compile.
    model.compile(optimizer='adam',loss='binary_crossentropy',metrics=['acc'])
    return model

estimators = []
estimators.append(('standardize', StandardScaler()))
estimators.append(('mlp', KerasClassifier(build_fn=create_smaller, epochs=100, batch_size=5, verbose=0)))
pipeline = Pipeline(estimators)
kfold = StratifiedKFold(n_splits=10, shuffle=True, random_state=seed)
results = cross_val_score(pipeline, X, encoded_Y, cv=kfold)
print("Smaller: %.2f%% (%.2f%%)" % (results.mean()*100, results.std()*100))

Smaller: 84.12% (8.72%)


In [18]:
# larger model
def create_larger():
	# create model
    model = Sequential()
    model.add(Dense(60,activation='relu',input_shape=(60,)))
    model.add(Dense(60,activation='relu'))
    model.add(Dense(30,activation='relu'))
    model.add(Dense(1,activation='sigmoid'))
	# Compile model
    model.compile(optimizer='adam',loss='binary_crossentropy',metrics=['acc'])
    return model
estimators = []
estimators.append(('standardize', StandardScaler()))
estimators.append(('mlp', KerasClassifier(build_fn=create_larger, epochs=100, batch_size=5, verbose=0)))
pipeline = Pipeline(estimators)
kfold = StratifiedKFold(n_splits=10, shuffle=True, random_state=seed)
results = cross_val_score(pipeline, X, encoded_Y, cv=kfold)
print("Larger: %.2f%% (%.2f%%)" % (results.mean()*100, results.std()*100))


Larger: 85.55% (7.13%)


## Step #5 -Overfitting our model.

In [19]:
# larger model
def create_larger():
	# create model
    model = Sequential()
    model.add(Dense(60,activation='relu',input_shape=(60,)))
    model.add(Dense(60,activation='relu'))
    model.add(Dense(30,activation='relu'))
    model.add(Dense(30,activation='relu'))
    model.add(Dense(30,activation='relu'))
    model.add(Dense(60,activation='relu'))
    model.add(Dense(1,activation='sigmoid'))
	# Compile model
    model.compile(optimizer='adam',loss='binary_crossentropy',metrics=['acc'])
    return model
estimators = []
estimators.append(('standardize', StandardScaler()))
estimators.append(('mlp', KerasClassifier(build_fn=create_larger, epochs=50, batch_size=5, verbose=0)))
pipeline = Pipeline(estimators)
kfold = StratifiedKFold(n_splits=10, shuffle=True, random_state=seed)
results = cross_val_score(pipeline, X, encoded_Y, cv=kfold)
# print("Larger - Result's Mean: %.2f" % result.mean()*100)
# print("Larger - Result's Standard Deviation: %.2f" % result.std()*100)
print("Larger: %.2f%% (%.2f%%)" % (results.mean()*100, results.std()*100))



Larger: 84.07% (6.95%)


## Step #6 - What was the best score?
It would be 'Smaller: 84.12% (6.99%)'.

## Step #7 - Rewriting the code using the Keras Functional API.

In [27]:
def keras_functional_api():
    inputs = tf.keras.Input(shape=(60,))
    x = Dense(60,activation='relu')(inputs)
    x = Dense(30,activation='relu')(x)
    outputs = Dense(1,activation='sigmoid')(x)
    model = tf.keras.Model(inputs,outputs)
    # Compile model.
    model.compile(optimizer='adam',loss='binary_crossentropy',metrics=['acc'])
    return model
model = keras_functional_api()
history = model.fit(X,encoded_Y,epochs=100, batch_size=5)
results = history.history['acc']
print("Functional API: Mean (%.2f%%) & Standard Deviation (%.2f%%)" % (np.mean(results)*100, np.std(results)*100))

Train on 208 samples
Epoch 1/100
208/208 [==============================] - 1s 7ms/sample - loss: 0.6886 - acc: 0.5192
Epoch 2/100
208/208 [==============================] - 0s 825us/sample - loss: 0.6478 - acc: 0.6394
Epoch 3/100
208/208 [==============================] - 0s 823us/sample - loss: 0.6125 - acc: 0.7212
Epoch 4/100
208/208 [==============================] - 0s 812us/sample - loss: 0.5652 - acc: 0.7452
Epoch 5/100
208/208 [==============================] - 0s 800us/sample - loss: 0.5253 - acc: 0.7837
Epoch 6/100
208/208 [==============================] - 0s 790us/sample - loss: 0.4818 - acc: 0.7644
Epoch 7/100
208/208 [==============================] - 0s 760us/sample - loss: 0.4462 - acc: 0.7981
Epoch 8/100
208/208 [==============================] - 0s 734us/sample - loss: 0.4284 - acc: 0.7933
Epoch 9/100
208/208 [==============================] - 0s 789us/sample - loss: 0.4095 - acc: 0.8173
Epoch 10/100
208/208 [==============================] - 0s 826us/sample - loss: 0

208/208 [==============================] - 0s 911us/sample - loss: 0.0088 - acc: 1.0000
Epoch 83/100
208/208 [==============================] - 0s 883us/sample - loss: 0.0086 - acc: 1.0000
Epoch 84/100
208/208 [==============================] - 0s 862us/sample - loss: 0.0087 - acc: 1.0000
Epoch 85/100
208/208 [==============================] - 0s 887us/sample - loss: 0.0087 - acc: 1.0000
Epoch 86/100
208/208 [==============================] - 0s 879us/sample - loss: 0.0084 - acc: 1.0000
Epoch 87/100
208/208 [==============================] - 0s 888us/sample - loss: 0.0074 - acc: 1.0000
Epoch 88/100
208/208 [==============================] - 0s 838us/sample - loss: 0.0073 - acc: 1.0000
Epoch 89/100
208/208 [==============================] - 0s 834us/sample - loss: 0.0069 - acc: 1.0000
Epoch 90/100
208/208 [==============================] - 0s 823us/sample - loss: 0.0068 - acc: 1.0000
Epoch 91/100
208/208 [==============================] - 0s 756us/sample - loss: 0.0063 - acc: 1.0000
Epo

## Step #8: Rewriting the code by doing Model Subclassing.

In [21]:
class MyModel(tf.keras.Model):
  def __init__(self):
    super(MyModel, self).__init__()
    self.dense1 = Dense(60, activation='relu')
    self.dense2 = Dense(30, activation='relu')
    self.dense3 = Dense(1, activation='sigmoid')

  def call(self, inputs):
    x = self.dense1(inputs)
    x = self.dense2(x)
    return self.dense3(x)

def subclass_model():
    inputs = tf.keras.Input(shape=(60,))
    mymodel = MyModel()
    outputs = mymodel.call(inputs)
    # Keras Model.
    model = tf.keras.Model(inputs,outputs)
    # Compile.
    model.compile(optimizer='adam',loss='binary_crossentropy',metrics=['acc'])
    return model
# Training model.
model = subclass_model()
model.fit(X,encoded_Y,epochs=100, batch_size=5)

Train on 208 samples
Epoch 1/100
208/208 [==============================] - 1s 6ms/sample - loss: 0.6603 - acc: 0.6106
Epoch 2/100
208/208 [==============================] - 0s 785us/sample - loss: 0.6152 - acc: 0.7212
Epoch 3/100
208/208 [==============================] - 0s 734us/sample - loss: 0.5815 - acc: 0.7163
Epoch 4/100
208/208 [==============================] - 0s 756us/sample - loss: 0.5275 - acc: 0.7644
Epoch 5/100
208/208 [==============================] - 0s 756us/sample - loss: 0.4940 - acc: 0.8029
Epoch 6/100
208/208 [==============================] - 0s 761us/sample - loss: 0.4603 - acc: 0.8077
Epoch 7/100
208/208 [==============================] - 0s 767us/sample - loss: 0.4551 - acc: 0.7837
Epoch 8/100
208/208 [==============================] - 0s 759us/sample - loss: 0.4096 - acc: 0.8462
Epoch 9/100
208/208 [==============================] - 0s 748us/sample - loss: 0.4123 - acc: 0.8077
Epoch 10/100
208/208 [==============================] - 0s 765us/sample - loss: 0

208/208 [==============================] - 0s 710us/sample - loss: 0.0084 - acc: 1.0000
Epoch 83/100
208/208 [==============================] - 0s 736us/sample - loss: 0.0096 - acc: 1.0000
Epoch 84/100
208/208 [==============================] - 0s 740us/sample - loss: 0.0087 - acc: 1.0000
Epoch 85/100
208/208 [==============================] - 0s 776us/sample - loss: 0.0072 - acc: 1.0000
Epoch 86/100
208/208 [==============================] - 0s 791us/sample - loss: 0.0085 - acc: 1.0000
Epoch 87/100
208/208 [==============================] - 0s 763us/sample - loss: 0.0073 - acc: 1.0000
Epoch 88/100
208/208 [==============================] - 0s 760us/sample - loss: 0.0068 - acc: 1.0000
Epoch 89/100
208/208 [==============================] - 0s 798us/sample - loss: 0.0060 - acc: 1.0000
Epoch 90/100
208/208 [==============================] - 0s 762us/sample - loss: 0.0062 - acc: 1.0000
Epoch 91/100
208/208 [==============================] - 0s 708us/sample - loss: 0.0057 - acc: 1.0000
Epo

## Step #9 - K-Fold Validation

In [22]:
k = 10
num_val_samples = len(X) // k
num_epochs = 100
all_scores = []
for i in range(k):
    print('processing fold #', i)
    val_data = X[i * num_val_samples: (i + 1) * num_val_samples]
    val_targets = encoded_Y[i * num_val_samples: (i + 1) * num_val_samples]
    partial_train_data = np.concatenate(
        [X[:i * num_val_samples],
        X[(i + 1) * num_val_samples:]],
        axis=0)
    partial_train_targets = np.concatenate(
        [encoded_Y[:i * num_val_samples],
        encoded_Y[(i + 1) * num_val_samples:]],
        axis=0)
    model = subclass_model()
    model.fit(partial_train_data, partial_train_targets,epochs=num_epochs, batch_size=1, verbose=0)
    val_mse, val_mae = model.evaluate(val_data, val_targets, verbose=0)
    # Evaluation.
    results = model.evaluate(val_data, val_targets, verbose=0)
    all_scores.append(val_mae)
#     all_scores = np.append(val_mae,results[1])
# k-fold scores.
print(all_scores)
print(f"K-Fold Mean Accuracy:{str(np.mean(all_scores)*100)} and Standard Deviation: {str(np.std(all_scores)*100)}")

processing fold # 0
processing fold # 1
processing fold # 2
processing fold # 3
processing fold # 4
processing fold # 5
processing fold # 6
processing fold # 7
processing fold # 8
processing fold # 9
[0.25, 0.65, 0.55, 0.85, 0.6, 0.25, 0.5, 0.65, 0.5, 1.0]
K-Fold Mean Accuracy:58.000004291534424 and Standard Deviation: 22.15851992368698


In [23]:
k = 10
num_val_samples = len(X) // k
# print(num_val_samples)
num_of_epochs = 100
all_scores = []
for i in range(k):
    print(f"Processes fold: {str(i)}")
    val_data = X[i * num_val_samples: (i + 1) * num_val_samples]

Processes fold: 0
Processes fold: 1
Processes fold: 2
Processes fold: 3
Processes fold: 4
Processes fold: 5
Processes fold: 6
Processes fold: 7
Processes fold: 8
Processes fold: 9
